### PR TITLE
feat(subagents): add unified capacity and durable batch execution

### LINKED ISSUE #4993: [feat] Deerflow对并发能力的硬编码限制，影响长程任务的执行，无法完成问题分析
### Before you start

- [x] I searched [existing issues](https://github.com/bytedance/deer-flow/issues?q=is%3Aissue) and this is not a duplicate.

### Problem / motivation

目前在使用deerflow解决业务问题的时候遇到这样一个场景：在24小时内，使用智能体分析用户提出的5000条需求项，并输出分析结果。
在实际测试中发现，每条需求项相互独立，如果多条需求项放在一起分析，由于上下文不相关内容干扰，会显著影响实际效果，因此会启动1个子智能体分析1条需求项，而5000条需求项，就需要启动5000次子智能体。
而使用deerflow开发的智能体会有如下2个问题：

**第一个问题：**
在deerflow中，针对一次run，使用参数max_total_per_run限制了子智能体的使用次数，并且启用了校验，超过50次就无法再次启动新的子智能体，这会导致在这个场景中，使用deerflow创建的智能体无法完成该业务场景下的分析。
这个限制体现的config.yaml配置文件中：
 `max_total_subagents`, clamped to the same 1-50 range.
 max_total_per_run: 6
在代码中：
MIN_TOTAL_SUBAGENTS_PER_RUN = 1
MAX_TOTAL_SUBAGENTS_PER_RUN = 50
无法通过配置文件改变这个限制的束缚，达到业务的目标。

**第二个问题：**
在deerflow中，子智能体的执行队列，使用代码硬性限制了只有3个worker可以同时运行：
 Thread pool for background task scheduling and orchestration
_scheduler_pool = ThreadPoolExecutor(max_workers=3, thread_name_prefix="subagent-scheduler-")
这严重限制了智能体的执行速度，我们实际测试40条需求项分析，启动4个task调用子智能体并发运行，需要耗时大约8个小时才能完成，远不能达成业务的目标，但是通过配置又无法解除限制，提升智能体的并发能力

研究了opencode和openclaw智能体的实现方式，它们要么没有限制，要么可以通过自定义配置的方式解除限制
| 维度                        | deerflow                                      | opencode                | openclaw                                                  |
| --------------------------- | --------------------------------------------- | ----------------------- | --------------------------------------------------------- |
| "一次 run 累计委派次数"上限 | ✅ 有：max_total_per_run 默认 6（clamp 1-50） | ❌ 无                   | ❌ 无                                                     |
| 记账方式                    | durable delegation ledger，按 run_id 计数     | 无任何度数              | 无累计记账                                                |
| 同时活动子代理上限          | 执行层线程池 3（硬编码）                      | 无                      | maxChildrenPerAgent=5（per-parent）+ Subagent lane 全局 8 |
| 嵌套深度                    | 无                                            | subagent_depth（默认1） | maxSpawnDepth（默认1）                                    |
| 子代理完成后是否释放配额    | 释放（但你 6 次就用完了）                     | —                       | ✅ 释放（可无限次 spawn）                                 |

### Proposed solution

能够通过自定义配置的方式，配置上述2个并发参数的值，提升deerflow智能体的并发能力

### Affected area(s)

Agents / LangGraph (graph, prompts, langgraph.json)

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AnnaSuSu**
Fixes #4993

Related to #3099, #2670, #1319, #1339, #3857, #4290, #4560, #3948, and #1223. This also carries forward the concurrency-alignment direction from #3415.

## Why

DeerFlow previously advertised a configurable per-response subagent concurrency while the native executor remained fixed at three process-local slots. Raising the ordinary per-run delegation limit would not make large fan-out durable or recoverable, and would weaken an important runaway guard.

This change aligns ordinary delegation with real process capacity and adds an explicit, opt-in batch path for large collections of independent items. A 5,000-item workload becomes representable, bounded, owner-scoped, and recoverable without pretending that 5,000 subagents can run simultaneously.

## What changed

- Added startup-only `subagent_runtime` configuration for the process-wide running limit, bounded queue, queue/reject policy, and queue timeout. Ordinary `task` calls and batch items share one async FIFO admission controller; waiting work no longer occupies scheduler threads.
- Resolved the ordinary `max_concurrent_subagents` once and reused it across the lead-agent prompt, middleware enforcement, client construction, and real executor capacity. Existing defaults remain three running subagents, six cumulative ordinary delegations, and a hard ordinary-run total of 50.
- Added an explicit `SubagentRuntime` dependency for direct `create_deerflow_agent(...)` integrations. Direct graphs can now share the same middleware limits, bound ordinary/batch tools, caller-owned `AppConfig`, real execution controller, and optional durable-worker lifecycle without the factory reading YAML or silently starting infrastructure.
- Added explicit `batch_task` submission. Batch mode is never inferred from prompt wording or item count, and is disabled by default.
- Added SQL-backed batches and items with separate total, live, and running limits, idempotent submission/item identities, leases, bounded retry, pause/resume/cancel, failed-item retry, result previews, and JSONL export. Execution is documented as at-least-once.
- Added owner- and thread-scoped Gateway APIs plus a feature-gated batch panel on default and custom-agent chat pages.
- Added migration `0016_subagent_batches` after `0015_scheduled_task_enqueue`, synchronized the Helm-embedded config schema version, and documented the full runtime, recovery, authorization, scaling, direct-factory, and OpenClaw comparison contracts.

## Surface area

- [x] Frontend UI
- [x] Backend API
- [x] Agents / LangGraph
- [ ] Sandbox
- [ ] Skills
- [ ] Dependencies
- [ ] Default behavior change
- [ ] Docs / tests / CI only

## Screenshots / Recording

image

## Validation

- `cd backend && make format` — passed
- `cd backend && make lint` — passed
- Original focused and integration backend capacity, direct-factory runtime/config binding, executor, middleware, prompt, batch tool/service/repository/API, migration, bootstrap, lifecycle, client, and guidance suite — 665 passed
- Review-affected backend capacity, prompt/middleware, client, executor, batch service/repository/API, guidance, and model-authorization regression set — 203 passed
- `cd backend && make test` — 12,239 passed, 76 skipped; the remaining 15 failures are the same DNS/SSRF baseline reproduced at the PR base (public test hosts resolve to a private address in this environment)
- `cd frontend && pnpm format` — passed
- `cd frontend && pnpm lint` — passed
- `cd frontend && pnpm typecheck` — passed
- `cd frontend && pnpm test` — 1,045 passed
- CI `Layer 2 — full-stack render (no API key)` — passed
- CI `e2e-tests` — passed
- All required PR checks — passed, including backend unit tests, frontend unit tests, both lints, E2E, both replay layers, blocking-I/O, install collection, and chart validation
- Helm lint, template render, sandbox Service-type assertions, and config-version drift check — passed
- `git diff --check origin/main...HEAD` — passed

## AI assistance

**Tool(s) used:** Codex

**How you used it:** Used Codex to inspect the existing delegation and persistence architecture, compare the explicit OpenClaw Swarm boundaries, implement the backend/frontend/tests/documentation, and run validation. The resulting changes and test outcomes were reviewed locally before submission.

- [x] I've read and understood the code changes in this PR and take full responsibility for them.

### CODE REVIEW

File: backend/app/gateway/routers/features.py [hunk 1]
```diff
@@ -59,4 +68,8 @@ async def list_features(request: Request, config: AppConfig = Depends(get_config
         # capability that actually started rather than a hot-reloaded config
         # value that would require a Gateway restart to take effect.
         mcp_tasks=McpTasksFeature(enabled=bool(getattr(request.app.state, "mcp_tasks_available", False))),
+        subagent_batches=SubagentBatchesFeature(
+            enabled=bool(getattr(request.app.state, "subagent_batches_available", False)),
+            max_running=configured_subagent_max_running(),
+        ),
```
[the remark this answers is not in the document]
Comment (AnnaSuSu) [reply backend/app/gateway/routers/features.py:82]: Fixed in 3960d6ec. The features endpoint now reports repository_available and worker_running independently; enabled remains a compatibility alias during rolling upgrades. Read, list, and export capability now follows repository availability, so persisted history is not hidden when the worker is stopped.

File: frontend/src/components/workspace/thread-subagent-batches.tsx [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+export function ThreadSubagentBatches({ threadId }: { threadId: string }) {
+  const { t } = useI18n();
+  const { enabled } = useSubagentBatchesEnabled();
+  const batchesQuery = useSubagentBatches(threadId, { enabled });
+  const control = useControlSubagentBatch(threadId);
+  const batches = batchesQuery.data ?? [];
+  const activeCount = batches.filter(isActiveSubagentBatch).length;
+
+  if (!enabled) return null;
+
```
[the remark this answers is not in the document]
Comment (AnnaSuSu) [reply frontend/src/components/workspace/thread-subagent-batches.tsx:60]: Fixed in 3960d6ec. The panel now queries when the repository is available, remains visible for threads with historical batches while the worker is stopped, disables worker-dependent controls and polling, and keeps item inspection and JSONL export available. Deployments with no worker and no history keep the unused trigger hidden. Covered by DOM tests.

File: frontend/src/components/workspace/thread-subagent-batches.tsx [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+              batch.max_running_items,
+            )}
+          </p>
+        </div>
+        <Badge variant="outline">{labels[batch.status]}</Badge>
+      </div>
+      <Progress
+        className="mt-3 h-1.5"
+        value={(completed / batch.total_items) * 100}
+      />
```
[the remark this answers is not in the document]
Comment (AnnaSuSu) [reply frontend/src/components/workspace/thread-subagent-batches.tsx]: Fixed in 3960d6ec. Progress is now normalized through a helper that returns 0 for non-positive or non-finite totals and clamps the result to 0-100. Added tests for zero totals and over-counted terminal items.

File: backend/packages/harness/deerflow/subagents/executor.py [hunk 4]
```diff
@@ -883,6 +893,27 @@ async def _build_initial_state(self, task: str) -> tuple[dict[str, Any], list[Ba
...
     async def _aexecute(self, task: str, result_holder: SubagentResult | None = None) -> SubagentResult:
+        """Execute after acquiring the process-wide native-subagent slot."""
+        result = result_holder
+        if result is None:
+            result = SubagentResult(
+                task_id=str(uuid.uuid4())[:8],
+                trace_id=self.trace_id,
+                status=SubagentStatus.PENDING,
+            )
+        try:
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/subagents/executor.py:908]: Suggestion: a `SubagentCapacityTimeout` raised while waiting for a shared process slot is terminalized as `FAILED`, and for batch items that consumes one of the item's `max_attempts` (the service finalizes with `succeeded=False`, so `finalize_item` requeues only while attempts remain). Since `SubagentBatchService.run_once` computes `available` from `max_running - len(self._executions)` without visibility into ordinary `task` calls that hold slots, a claimed item can sit in the capacity queue behind ordinary work for > `queue_timeout_seconds` (default 300s) on *every* attempt and permanently fail with attempts exhausted without ever executing. Consider distinguishing admission failures (`SubagentCapacityRejected`/`SubagentCapacityTimeout`) from execution failures in `SubagentResult` so `_execute_item` can requeue without burning an attempt - an item that never ran is a scheduling condition, not an execution failure.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+    def _batch_dict(row: SubagentBatchRow) -> dict[str, Any]:
+        data = row.to_dict()
+        for key in _BATCH_TIMESTAMP_FIELDS:
+            if data.get(key) is not None:
+                data[key] = coerce_iso(data[key])
+        return data
+
+    @staticmethod
+    def _item_dict(row: SubagentBatchItemRow) -> dict[str, Any]:
+        data = row.to_dict(exclude={"prompt"})
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Suggestion: `_item_dict` excludes `prompt` but keeps the full `result` text (up to `max_result_chars` = 100,000 chars per item). `GET .../items` accepts `limit=500`, so a single page can carry ~50MB of result text that no consumer renders - the batch panel uses only `result_preview`, and the documented path for full results is the JSONL export. Consider excluding `result` from `list_items` the same way `prompt` is excluded (or gating it behind an explicit `include_result` flag) so list/detail traffic stays bounded.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                await self._refresh_batch_status(session, batch, now=completed_at)
+            await session.commit()
+            return True
+
+    async def _refresh_batch_status(self, session: AsyncSession, batch: SubagentBatchRow, *, now: datetime) -> None:
+        counts = await self._counts(session, batch.id)
+        terminal = sum(counts[state] for state in ITEM_TERMINAL_STATUSES)
+        if terminal >= batch.total_items:
+            if batch.status != "cancelled":
+                batch.status = "completed"
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Nit: batch-level `failed` exists in `BATCH_TERMINAL_STATUSES` and in the frontend status labels, but nothing ever writes it - `_refresh_batch_status` marks the batch `completed` whenever all items are terminal, even if every single item failed. Either derive a terminal `failed` status (e.g. succeeded == 0 and failed > 0) or drop the unreachable state so the UI label isn't dead code.

**Review by willem-bd — COMMENTED** [review]
Reviewed at head 3960d6ec. The overall architecture is solid: one startup-scoped admission controller shared by ordinary `task` and batch items, explicit-only batch mode, lease-based multi-worker recovery with stable item identity and idempotent submission, and a clean owner-scoped API with repository/worker capabilities split so history survives a stopped worker. The three earlier Copilot findings (features `enabled` alias, hidden history panel, unguarded progress division) are all verified fixed at this head. Docs/AGENTS.md/config-version/Helm sync is thorough, and the capacity controller's timeout/cancel slot-transfer edge cases are carefully handled with tests. Three new findings below, all suggestion/nit level - nothing blocking.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+                                SubagentBatchItemRow.status.not_in(ITEM_TERMINAL_STATUSES),
+                            )
+                            .with_for_update()
+                        )
+                    ).scalars()
+                )
+                for item in items:
+                    item.cancel_requested_at = now
+                    item.updated_at = now
+                    if item.status not in ("leased", "running"):
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: [P1] Recover cancelled in-flight items after worker loss

Setting the batch to `cancelled` removes it from the `claim_items()` scan, which only considers queued/running batches, but this branch deliberately leaves leased/running rows nonterminal. If the worker disappears after this commit, lease expiration is never swept and the item remains `running` forever even though its batch is terminal. Please either transition active rows to cancelled here (stale finalization will already be fenced by status) or add recovery of expired active rows in cancelled batches.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+class SubagentBatchRepository:
+    """Durable batch/item state with lease-based multi-worker claiming."""
+
+    def __init__(self, session_factory: async_sessionmaker[AsyncSession]) -> None:
+        self._sf = session_factory
+
+    @staticmethod
+    def _batch_dict(row: SubagentBatchRow) -> dict[str, Any]:
+        data = row.to_dict()
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: [P1] Keep the persisted execution snapshot out of public responses

`row.to_dict()` includes `execution_spec`, and `batch_task` persists the subagent configuration—including its system prompt—plus user-role, OAuth/channel identity, internal-call, and authorization attributes there. The owner-facing list/status routes return this dictionary directly, so a user allowed to invoke an operator-managed subagent can retrieve that hidden prompt and server-owned auth context. Please use separate internal and public projections and exclude `execution_spec` from API/tool status responses.

File: backend/packages/harness/deerflow/agents/lead_agent/agent.py [hunk 9]
```diff
@@ -870,7 +876,10 @@ def _assemble_lead_agent(config: RunnableConfig, *, app_config: AppConfig) -> Le
     requested_model_name: str | None = cfg.get("model_name") or cfg.get("model")
     is_plan_mode = cfg.get("is_plan_mode", False)
     requested_subagent_enabled = cfg.get("subagent_enabled", False)
-    max_concurrent_subagents = cfg.get("max_concurrent_subagents", 3)
+    max_concurrent_subagents = effective_subagent_concurrency(
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/agents/lead_agent/agent.py:885]: [P2] Use the startup-frozen process capacity here

The Gateway injects the live `get_config()` result on every run, while the shared execution controller is configured only once from startup configuration. After a config hot reload, this calculation can therefore use a different `max_running` value from the actual executor (for example, I reproduced prompt/middleware capacity 12 with an executor frozen at 3). That violates the single-capacity invariant and makes the startup-only boundary misleading. Please derive the global/Gateway value from the configured process capacity, or inject the startup runtime snapshot separately.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 5 above)
Comment (willem-bd) [inline backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: [P2] Omit full results from the batch item list

Excluding only `prompt` still serializes the full `result`. The panel fetches up to 100 items and renders only `result_preview`, while each result may be configured as large as one million characters, so merely opening the panel can transfer and retain tens of megabytes. Please use a list projection that excludes `result` and reserve the full payload for a detail/export path (ideally streamed or paged without materializing a large result set).

**Review by willem-bd — COMMENTED** [review]
Found four new issues in the durable batch execution implementation.

**Review by willem-bd — CHANGES_REQUESTED** [review]
Requesting changes on `734df7c9` for the verified inline findings:

1. **[P1] Cancelled in-flight items are not recovered after worker loss.** A cancelled batch is excluded from the recovery scan while its leased/running items remain nonterminal, so an expired item can stay `running` forever. Reproduced with a SQLite regression probe: https://github.com/bytedance/deer-flow/pull/4998#discussion_r3844176872
2. **[P1] Internal execution state is exposed by owner-facing APIs.** The raw batch projection includes `execution_spec`, including the subagent system prompt and server-owned authorization context: https://github.com/bytedance/deer-flow/pull/4998#discussion_r3844176877
3. **[P2] Prompt/middleware concurrency can drift from the startup-frozen executor capacity after config reload:** https://github.com/bytedance/deer-flow/pull/4998#discussion_r3844176882
4. **[P2] The polling item list returns full stored results even though the UI renders only previews:** https://github.com/bytedance/deer-flow/pull/4998#discussion_r3844176887
5. **[P2] The large-batch UI hard-codes `limit=100` with no offset, status filter, or load-more path.** For the advertised 5,000-item workload, failures after position 99 are invisible and cannot be retried through the panel. Please add pagination/load-more or a failed-items view.

The current CI is green. Focused validation passed 32 backend tests and 15 frontend tests; the additional cancellation recovery probe fails as described above.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 5 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Implemented in 4f130ca9. Item listing now uses an explicit public projection and excludes the full result by default. The repository has an explicit include_result opt-in used only by the paged JSONL export, while the polling/list path returns the bounded preview. Added repository and router regressions.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 8 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Implemented in 4f130ca9. Public batch and item responses now use explicit allowlisted projections. execution_spec, user/submission/run/tool-call identity, prompts, authorization context, lease internals, and full result text are excluded; the worker claim path has a separate internal execution projection. Added projection regressions with a private system prompt and authz attributes.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 5 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Implemented in 4f130ca9. The owner-facing item projection omits full result text and the frontend item type no longer declares it. Only the paged JSONL export opts into full stored results. Repository/router tests verify both the bounded list response and full export behavior.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 6 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Implemented in 4f130ca9. When every item is terminal, the batch now becomes failed if at least one item failed and none succeeded; mixed success/failure remains completed with item-level counts. Added an all-failed repository regression.

File: backend/packages/harness/deerflow/persistence/subagent_batches/sql.py (hunk 7 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/persistence/subagent_batches/sql.py]: Implemented in 4f130ca9. Batch cancellation now atomically terminalizes every nonterminal item, clears its lease owner/expiry, and records cancellation. Late worker finalization is fenced because finalize_item only accepts leased/running rows. Added a regression that cancels a running item and verifies the stale completion is rejected.

File: backend/packages/harness/deerflow/subagents/executor.py (hunk 4 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/subagents/executor.py:908]: Implemented in 4f130ca9. Capacity rejection and queue timeout are now tagged as pre-execution admission failures on SubagentResult. The batch service routes those results to a repository transition that releases the lease and restores the claim attempt, so no retry budget is consumed before model execution starts. Added executor, service, and SQL repository regressions for this path.

File: backend/packages/harness/deerflow/agents/lead_agent/agent.py (hunk 9 above)
Comment (AnnaSuSu) [reply backend/packages/harness/deerflow/agents/lead_agent/agent.py:885]: Implemented in 4f130ca9. Gateway lead assembly now reads configured_subagent_max_running() from the startup-frozen controller snapshot and passes that value through initial resolution, SubagentLimitMiddleware construction, and prompt rendering. DeerFlowClient stores the same startup snapshot. Regressions cover a live config reloaded to 12 while execution remains frozen at 3.

**Comment by AnnaSuSu** [comment]
@willem-bd I pushed the review fixes in 4f130ca9 and replied on each inline thread.

I also addressed the non-inline frontend pagination finding:
- batch items now use offset-based infinite pagination with an explicit Load more action;
- the API client supports offset, limit, and status;
- once more than one page is loaded, automatic 3-second all-page refetching stops so polling remains bounded for large batches.

Validation on the updated head:
- review-affected backend regression set: 203 passed;
- local full backend suite: 12,239 passed, 76 skipped, with only the same 15 DNS/SSRF environment baseline failures;
- frontend check and all 1,045 unit tests passed;
- every required PR check now passes, including backend-unit-tests, e2e-tests, frontend-unit-tests, both lints, and both replay layers.

I left the human review threads unresolved for your re-review.

**Review by WillemJiang — APPROVED** [review]
